src/modules/recursive_partitioning/DT_impl.hpp

```diff
@@ -486,8 +485,18 @@ DecisionTree<Container>::expand(const Accumulator &state,
             Index stats_i = static_cast<Index>(state.stats_lookup(i));
             assert(stats_i >= 0);
 
-            // 1. Set the prediction for current node from stats of all rows
-            predictions.row(current) = state.node_stats.row(stats_i);
+            if (statCount(predictions.row(current)) !=
+                    statCount(state.node_stats.row(stats_i))){
+                // Predictions for each node is set by its parent using stats
+                // recorded while training parent node. These stats do not include
+                // rows that had a NULL value for the primary split feature.
+                // The NULL count is included in the 'node_stats' while training
+                // current node. Further, presence of NULL rows indicate that
+                // stats used for deciding 'children_wont_split' are inaccurate.
+                // Hence avoid using the flag to decide termination.
+                predictions.row(current) = state.node_stats.row(stats_i);
+                children_wont_split = false;
+            }
```

Comment:
- `children_wont_split` is **one** of the factors that determines if training should stop after current iteration. `children_wont_split=true` implies training stops; `children_wont_split=false` implies other flags determine termination. 
- The lines 516-547 are finding the best feature to split on and are necessary - independent of `children_wont_split` and independent of the result of line 490. 

I could exchange sections 1 and 2 since they're independent, if that helps in reading the code.